### PR TITLE
 Gives the previous cursor in the scroll block

### PR DESCRIPTION
**Pull request by GCorbel**
In addition to the next cursor, blocks in `Mongo::Scrollable#scroll` and `Mongoid::Criteria::Scrollable` gives an additional cursor with allow to fetch records before the first one returned.

The previous cursor is created based on the first record with a `previous` flag set to `true`. For example, if we have `Feed::Item` with a name from 0 to 3, we have this behavior :

```
cursor = nil
first_previous_cursor = nil

Feed::Item.asc(:name).limit(2).scroll do |record, next_cursor|
  puts record.name # => 0, 1
  cursor = next_cursor
end

Feed::Item.asc(:name).limit(2).scroll(cursor) do |record, _, previous_record|
  puts record.name # => 2, 3
  first_previous_cursor = previous_cursor
end

Feed::Item.asc(:name).limit(2).scroll(first_previous_cursor) do |record|
  puts record.name # => 0, 1
  second_previous_cursor = previous_cursor
end
```

### CODE REVIEW

File: lib/mongo/scrollable.rb [hunk 1]
```diff
@@ -16,21 +16,37 @@ def scroll(cursor_or_type = nil, options = nil, &_block)
...
+      else
+        # make a view
+        records = Mongo::Collection::View.new(
+          view.collection,
+          view.selector.merge(cursor.criteria),
+          sort: (view.sort || {}).merge(_id: scroll_direction),
+          skip: skip,
+          limit: limit
+        )
+      end
```
Comment (GCorbel) [inline lib/mongo/scrollable.rb:40]: This : 

1. Select records before the limit in the criteria
2. Reverse the sort
3. Take only the number of records in `limit`
4. Resort the order

Given that, if we have this list : 

```
1
2
3
4
5 # <= Cursor here
6 
```

We want to have 2 records before the cursor and keep the order. Returned values have to be `3, 4`. 


To do that, we first select only records be the cursor :

```
1
2
3
4
```

We reverse the selection :  

```
4
3
2
1
```

Take only the two first values : 

```
4
3
```

And finally revert the sorting : 

```
3
4
```

To avoid using aggregation, another solution was possible by sorting first with mongo and the reverse the records to display in the correct order
Another possible solution was to do the same in pure Ruby : 

```
view.sort[scroll_field] = -scroll_direction if cursor.previous
records = Mongo::Collection::View.new(
  view.collection,
  view.selector.merge(cursor.criteria),
  sort: (view.sort || {}).merge(_id: -scroll_direction),
  skip: skip,
  limit: limit
)
records.to_a.sort { |h1, h2| h1[scroll_field] <=> h2[scroll_field] }.each do |record|
  # ...
end
```

Even if this version is simpler, I chose to use aggregates for performance reason.

File: spec/mongoid/base64_encoded_cursor_spec.rb [hunk 2]
```diff
@@ -3,17 +3,18 @@
 describe Mongoid::Scroll::Base64EncodedCursor do
   context 'new' do
     context 'an empty cursor' do
-      let(:base64_string) { 'eyJ2YWx1ZSI6bnVsbCwiZmllbGRfdHlwZSI6IlN0cmluZyIsImZpZWxkX25hbWUiOiJhX3N0cmluZyIsImRpcmVjdGlvbiI6MSwiaW5jbHVkZV9jdXJyZW50IjpmYWxzZSwidGllYnJlYWtfaWQiOm51bGx9' }
+      let(:base64_string) { 'eyJ2YWx1ZSI6bnVsbCwiZmllbGRfdHlwZSI6IlN0cmluZyIsImZpZWxkX25hbWUiOiJhX3N0cmluZyIsImRpcmVjdGlvbiI6MSwiaW5jbHVkZV9jdXJyZW50IjpmYWxzZSwidGllYnJlYWtfaWQiOm51bGwsInByZXZpb3VzIjpmYWxzZX0=' }
```
Comment (GCorbel) [inline spec/mongoid/base64_encoded_cursor_spec.rb:6]: tokens changed because previous: false is encoded.

File: lib/mongoid/scroll/base_cursor.rb [hunk 3]
```diff
@@ -1,7 +1,7 @@
 module Mongoid
   module Scroll
     class BaseCursor
-      attr_accessor :value, :tiebreak_id, :field_type, :field_name, :direction, :include_current
+      attr_accessor :value, :tiebreak_id, :field_type, :field_name, :direction, :include_current, :previous
```
Comment (GCorbel) [inline lib/mongoid/scroll/base_cursor.rb:4]: I'm not sure about the naming of the previous flag.

**Review by GCorbel — COMMENTED** [review]
This is far from being perfect, but it is backward compatible and works for the case I need.

Improvements can be done, as give an hash in the block as :

```

Feed::Item.asc(:name).limit(2).scroll do |record, meta_data|
  meta_data[:next_cursor]
  meta_data[:previous_cursor]
  meta_data[:total]
  # ...
end
```

But it is not backward compatible and needs more work.

**Comment by dblock** [comment]
Impressive that you got it working :)

Let's start by adding to README? I'll do a thorough review when I get a moment.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 ### 1.0.2 (Next)
```
Comment (dblock) [inline CHANGELOG.md:1]: Since we're adding a major feature, increment the version to `1.1.0` here and in `version.rb`.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
...
 
-* Your contribution here.
```
Comment (dblock) [inline CHANGELOG.md:3]: Put back this line below your entry.

File: README.md [hunk 6]
```diff
@@ -79,9 +79,17 @@ end
...
 ```ruby
-Feed::Item.desc(:position).limit(5).scroll(saved_cursor) do |record, next_cursor|
+Feed::Item.desc(:position).limit(5).scroll(saved_cursor) do |record, _, previous_cursor|
   # each record, one-by-one
-  saved_cursor = next_cursor
+  saved_cursor = previous_cursor
+end
+```
+
+A cursor to get the previous records is available.
```
Comment (dblock) [inline README.md:88]: The example below does not show the use of the previous cursor.

File: lib/mongo/scrollable.rb [hunk 7]
```diff
@@ -16,21 +16,37 @@ def scroll(cursor_or_type = nil, options = nil, &_block)
...
-        view.collection,
-        view.selector.merge(cursor.criteria),
-        sort: (view.sort || {}).merge(_id: scroll_direction),
-        skip: skip,
-        limit: limit
-      )
+
+      records = nil
+      if cursor.previous && limit
+        # scroll backards by reversing the sort order, limit and then reverse again
```
Comment (dblock) [inline lib/mongo/scrollable.rb:22]: typo: backwards

**Review by dblock — CHANGES_REQUESTED** [review]
I took a detailed look. I could be missing something but I think this is a different feature of scrolling backwards. I understand how that yields the previous cursor, but I believe this is more complicated than necessary. I am not sure of the value of scrolling backwards by making inverse queries, a caller can simply inverse the order by which to scroll.

So, a super inefficient and naive scroll backwards reaches the end saving the list of cursors, then yielding them in reverse order. In some way we want that for the previous cursor minus the inefficiencies of making all the queries.

A possibly much simpler implementation:

1. Start scroll, the starting cursor is `nil`.
2. Every page we may have the `current` cursor that we unpack and make a view with, and generate a `cursor` (aka a link to the next page).
3. Include the information about the `current` cursor in the generated next page one.
4. Add a property to the unpacked cursor/page called `previous` that returns (3), now you can make queries with it.

Basically, instead of `previous` being a boolean, make it the actual previous cursor "as is".

WDYT?

**Comment by GCorbel** [comment]
I can explain what I need more in details. 

ATM, In an API payload, we return the hash of a cursor as :

```
{
    "records": [
       # ...
    ],
    "paging": {
        "pageToken": "eyJ2YWx1ZSI6MTQwNTA4MTQ4Ni4yNDUsImZpZWxkX3R5cGUiOiJUaW1lIiwiZmllbGRfbmFtZSI6ImNyZWF0ZWRfYXQiLCJkaXJlY3Rpb24iOjEsImluY2x1ZGVfY3VycmVudCI6dHJ1ZSwidGllYnJlYWtfaWQiOiI1M2JmZDc4ZTdhYTU4ZDlkOWUwMDAwYjYifQ==",
        "perPage": 2,
        "nextPageToken": "eyJ2YWx1ZSI6MTQwNTA4Mzc2Ni44MzQsImZpZWxkX3R5cGUiOiJUaW1lIiwiZmllbGRfbmFtZSI6ImNyZWF0ZWRfYXQiLCJkaXJlY3Rpb24iOjEsImluY2x1ZGVfY3VycmVudCI6ZmFsc2UsInRpZWJyZWFrX2lkIjoiNTNiZmUwNzYwZjRjYTE5MWZiMDAwMDQ4In0="
    }
}
```

We can take the hash in `nextPageToken` and add it in the request parameter as `https://test.com?pageToken=eyJ2YWx1ZSI6MTQwNTA4Mzc2Ni44MzQsImZpZWxkX3R5cGUiOiJUaW1lIiwiZmllbGRfbmFtZSI6ImNyZWF0ZWRfYXQiLCJkaXJlY3Rpb24iOjEsImluY2x1ZGVfY3VycmVudCI6ZmFsc2UsInRpZWJyZWFrX2lkIjoiNTNiZmUwNzYwZjRjYTE5MWZiMDAwMDQ4In0=`

Then we take this param as : 

```
cursor = ::Mongoid::Scroll::Base64EncodedCursor.new(params["page_token"])
criteria.limit(limit).scroll(cursor) do |record, next_cursor|
   # ....
end
```

My need is to be able to do the same, but with the records before the records listed.

Given that, as we don't loop over each record, your proposal does not work.

File: README.md (hunk 6 above)
Comment (GCorbel) [reply README.md]: I tried to clarify the readme.

File: lib/mongo/scrollable.rb (hunk 7 above)
Comment (GCorbel) [reply lib/mongo/scrollable.rb]: I fixed the typo. Thanks.

File: CHANGELOG.md (hunk 4 above)
Comment (GCorbel) [reply CHANGELOG.md]: I did.

**Comment by dblock** [comment]
@GCorbel I'm a little slow on this one, so please bear with me.

I think I understand what you're trying to do. First, question: can one call your API and land randomly somewhere in the records or does it always require a cursor to land in the middle of data, except for page 1 where the cursor is `nil`? 

If you don't want to start a cursor in a random position, I think you don't need all this previous logic. Embed the previous cursor in every cursor in this library.

- Your `/api` yields currently `page2-cursor`, instead yield a pair `(page2-cursor, nil)`.
- Your `/api?cursor=(page2-cursor, nil)` currently yields `page3-cursor`, instead yield `(page3-cursor,page2-cursor)`.
- etc.

What I am proposing above is to change `cursor` to (optionally) store its `previous` cursor, not to burden your API to return and remember it.

Now, if you do want to produce a cursor from a random position in data (one that wasn't reached with the above iteration), so then your solution of reversing the order, making a database query, fetching the new position, re-generating a cursor in the `scroll` API works. But that's obviously very inefficient, so let's make sure we need this.

File: lib/mongoid/scroll/base_cursor.rb (hunk 3 above)
Comment (dblock) [reply lib/mongoid/scroll/base_cursor.rb:4]: Per my comment in https://github.com/mongoid/mongoid-scroll/pull/38#issuecomment-2311028872, this could be the cursor itself, `previous_cursor` and it would be deserialized into a `BaseCursor`.

**Comment by dblock** [comment]
.... actually now that I think about it more, my suggestion might not work - we'd need to store the entire history of cursors for every cursor, and that's going to be ... a lot 🤦‍♂️

**Comment by dblock** [comment]
So, a possibly simpler idea - `cursor.previous` would do the aggregation and find the previous cursor? This way we don't need to yield anything or add another option?

**Comment by GCorbel** [comment]
ok so, to be consistent, we should also have `cursor.next` ?

**Comment by GCorbel** [comment]
> cursor.previous would do the aggregation and return the previous cursor?

It means we will have an extra query each time, right ?

**Comment by GCorbel** [comment]
I tried very rapidly to implement a method `Mongoid::Scroll::Cursor#previous` that build a cursor base on the first record returned with the aggregation and wasn't able as we need `Mongoid::Criteria#options` and `Mongoid::Criteria#view` that are not available in instances of `Mongoid::Scroll::Cursor`. We need to alter the implementation to make it works.

Anyway, as it require an extra call each time to use `Mongoid::Criteria::Scrollable#scroll`, I'm not sure if it's a good idea.

Let me know what you think.

File: README.md [hunk 8]
```diff
@@ -76,7 +76,16 @@ Feed::Item.desc(:position).limit(5).scroll do |record, next_cursor|
...
 ```
 
-Resume iterating using the previously saved cursor.
+Resume iterating using saved cursor and save the cursor to go backward.
+
+```ruby
+Feed::Item.desc(:position).limit(5).scroll(saved_cursor) do |record, _, previous_cursor|
+  # each record, one-by-one
+  saved_cursor = previous_cursor
+end
```
Comment (dblock) [inline README.md:85]: We now yield the next and previous cursor every time, so everywhere we say `Feed::Item.desc(:position).limit(5).scroll do |record, next_cursor|` it should be `Feed::Item.desc(:position).limit(5).scroll do |record, next_cursor, previous_cursor|`.

I'd rename `saved_cursor` to `saved_next_cursor` and `saved_previous_cursor` to make it a bit clearer.

**Review by dblock — CHANGES_REQUESTED** [review]
Ok, I think you had it right the first time, your implementation is better. Thanks for hanging in here with me.

You're right to ask about naming `previous`. It's confusing because it's about the nature of the cursor. How about we introduce a `type` and give it a number (0 for next, 1 for previous)? Who knows what kind of other cursors we'll come up with :)

**Comment by GCorbel** [comment]
> How about we introduce a type and give it a number (0 for next, 1 for previous)? 

I will try with `type` that can be `:next` or `:previous`. It may be `:first`, `:current`, etc.

> Who knows what kind of other cursors we'll come up with :)

I may need to have a "firstPageToken" so I may add a "first_cursor". That's for another PR.

File: README.md (hunk 8 above)
Comment (GCorbel) [reply README.md:85]: WDYT about the fact that `next_cursor` is before `previous_cursor` ? If we can break things, I suggest having `cursors` which is a hash or object with different cursors so it will give : 

```
Feed::Item.desc(:position).limit(5).scroll(saved_cursor) do |record, cursors|
  # each record, one-by-one
  saved_cursors = cursors[:next]
end

saved_cursors[:next] # => next cursor
saved_cursors[:previous] # => previous cursor 
``` 

IMO, this is better, but it's not backward compatible.

Comment (dblock) [reply README.md:87]: A backwards compatible version could be `|record, next_cursor, iterator|` where `iterator` is a class that has `.next_cursor` and `.previous_cursor` and whatever else we choose to add in the future.

It's kinda ugly, so I would also be OK with breaking backwards compatibility, but I do think we should return an object and not just a hash of cursors.

1. Return an object instead of a cursors hash.
2. Increment version to 2.0 in this PR.
3. Add a section to https://github.com/mongoid/mongoid-scroll/blob/master/UPGRADING.md.

File: lib/mongo/scrollable.rb [hunk 9]
```diff
@@ -16,21 +16,37 @@ def scroll(cursor_or_type = nil, options = nil, &_block)
...
-      view = Mongo::Collection::View.new(
-        view.collection,
-        view.selector.merge(cursor.criteria),
-        sort: (view.sort || {}).merge(_id: scroll_direction),
-        skip: skip,
-        limit: limit
-      )
+
+      records = nil
+      if cursor.type == :previous && limit
```
Comment (dblock) [inline lib/mongo/scrollable.rb:21]: The `&& limit` is suspect. I think it's incorrect if `cursor.type == :previous` but `limit = nil`.

IMO `cursor.type` should always be assigned to either `:previous` or `:next`. Throw an exception inside deserialization when its neither of those values.

Here:

```
case cursor.type:
when :next, nil: ...
when :previous ...
else
   raise an error here
```

Comment (GCorbel) [reply lib/mongo/scrollable.rb]: I removed the `&& limit`. It wasn't needed anymore.

I still have to raise an error when there is no `type` options.

File: README.md (hunk 8 above)
Comment (GCorbel) [reply README.md:86]: I did changes.

File: README.md [hunk 10]
```diff
@@ -69,27 +69,38 @@ end
...
+  # each record, one-by-one
+  saved_iterator = iterator
+end
+```
+
+Loop over the first records again.
 
 ```ruby
-Feed::Item.desc(:position).limit(5).scroll(saved_cursor) do |record, next_cursor|
+Feed::Item.desc(:position).limit(5).scroll(saved_previous_cursor) do |record, iterator|
```
Comment (dblock) [inline README.md:92]:
```suggestion
Feed::Item.desc(:position).limit(5).scroll(saved_iterator.previous_cursor) do |record, iterator|
```

File: README.md [hunk 11]
```diff
@@ -98,19 +109,19 @@ end
...
-  saved_cursor = next_cursor
+  saved_iterator = iterator
 end
 ```
 
 Resume iterating using the previously saved cursor.
 
 ```ruby
-session[:feed_items].find.sort(position: -1).limit(5).scroll(saved_cursor, { field_type: DateTime }) do |record, next_cursor|
+session[:feed_items].find.sort(position: -1).limit(5).scroll(iterator.next_cursor, { field_type: DateTime }) do |record, iterator|
```
Comment (dblock) [inline README.md:122]:
```suggestion
session[:feed_items].find.sort(position: -1).limit(5).scroll(saved_iterator.next_cursor, { field_type: DateTime }) do |record, iterator|
```

File: UPGRADING.md [hunk 12]
```diff
@@ -1,5 +1,25 @@
 # Upgrading
 
+## Upgrading to >= 2.0.0
+
+The second argument passed to the block in `Mongoid::Criteria::Scrollable#scroll` and `Mongo::Scrollable#scroll` has changed. It is now an instance of `Mongoid::Criteria::Scrollable` that provides two methods: `next_cursor` and `previous_cursor`. The `next_cursor` method returns the same cursor as in versions prior to 2.0.0.
```
Comment (dblock) [inline UPGRADING.md:5]:
```suggestion
The second argument yielded in the block in `Mongoid::Criteria::Scrollable#scroll` and `Mongo::Scrollable#scroll` has changed from a cursor to an instance of `Mongoid::Criteria::Scrollable` which provides `next_cursor` and `previous_cursor`. The `next_cursor` method returns the same cursor as in versions prior to 2.0.0.
```

**Review by dblock — CHANGES_REQUESTED** [review]
Looks great. Some typo-things below.

File: lib/mongo/scrollable.rb (hunk 9 above)
Comment (GCorbel) [reply lib/mongo/scrollable.rb]: I added an error when the type is other than :next of :previous.

**Comment by GCorbel** [comment]
Typos are fixed. Thanks.

**Comment by dblock** [comment]
Merged, 👏 

Now that you wrote half of this library, would you like to join in as a (co)maintainer? Maybe make the next release?

If yes, drop me a note to dblock[at]dblock[dot]org with your rugygems username.

**Comment by GCorbel** [comment]
Thanks for the merge. I enjoy working with you.

I'm not very interested in being a maintainer, unless you don't want to be anymore.

**Comment by dblock** [comment]
> I'm not very interested in being a maintainer, unless you don't want to be anymore.

Too bad :( I don't mind, just trying to reduce bus factor. Appreciate any help you can give!
